Settings screen

TobiA34 opened the pull request:
- Add tabBar in the scene delegate
- Added gundam image
- Added sound kit sound to the project
- Created an audio manager
- Created settings screen


<img width="477" alt="image" src="https://github.com/TobiA34/Anime-API/assets/36420903/4d94a6b4-32b5-4cd1-ac4f-9cf93cdbfc90">

<img width="492" alt="image" src="https://github.com/TobiA34/Anime-API/assets/36420903/96c77d9a-33c9-4067-9ee5-267aad7b50bb">

TobiA34 committed:
- Add tabBar in scene delegate
- Added gundam image
- Added soundkit sound to project
- Created an audio manager
- Created settings screen

tunds commented on Anime-App/Feature/Anime/AudioManger/AudioManager.swift:
```diff
@@ -0,0 +1,41 @@
+    func stopSound() {
+        player?.stop()
+    }
+    
```
Clean up the white space

tunds commented on Anime-App/Feature/Anime/AudioManger/AudioManager.swift:
```diff
@@ -0,0 +1,41 @@
+        }
+    }
+    
+    func stopSound() {
```
I would just call this stop, you don't need stop sound cos you know this will stop the sound already

tunds commented on Anime-App/Feature/Anime/AudioManger/AudioManager.swift:
```diff
@@ -0,0 +1,41 @@
+        self.name = name
+    }
+    
+    func playSound() {
```
Same here lets call it play since we know already it will play a sound

tunds commented on Anime-App/Feature/Anime/AudioManger/AudioManager.swift:
```diff
@@ -0,0 +1,41 @@
+    private var name: String
+    private var player: AVAudioPlayer?
+
+    init(name: String) {
```
So i would actually move this param to the play function, think about it if you want to play different sounds you'd have to create a new instance of this everytime. Instead if it was a function param then you could just call `play(name: "pop")` and then just play random sounds without needing to create a new instance. One of those things to think about when you architect software is how do i want to be able to use this?

tunds commented on Anime-App/Feature/Anime/View/SettingsViewController.swift:
```diff
@@ -0,0 +1,179 @@
+        } else {
+            UserDefaults.standard.set(false, forKey: "UseHaptic")
+            guard let appDomain = Bundle.main.bundleIdentifier else { return }
+            UserDefaults.standard.removePersistentDomain(forName: appDomain)
```
So you don't want to do this where you remove it from user defaults you just want to set the the flag in user defaults to true or false. Also you could do it in one like this too.

```

 defaults.set(hapticFeedbackSwitch.isOn, forKey: "UseHaptic")

```

tunds commented on Anime-App/Feature/Anime/View/SettingsViewController.swift:
```diff
@@ -0,0 +1,179 @@
+    @objc func toggleHapticFeedback() {
+        if hapticFeedbackSwitch.isOn == true {
+            print("haptic Feedback is on")
+            defaults.set(true, forKey: "UseHaptic")
```
When naming keys in user defaults. it's better to go along the lines of `isHapticsEnabled` it's a lot clearer. Can you rename all of your keys to follow this pattern and can you use an enum to organise them all too.

tunds commented on Anime-App/Feature/Anime/View/SettingsViewController.swift:
```diff
@@ -0,0 +1,179 @@
+        if audioSwitch.isOn == true {
+            let audioIsOn = audioSwitch.isOn == true
+            print("Audio is on")
+            defaults.set(audioIsOn, forKey: "UseAudio")
```
Same comment as above

tunds commented on Anime-App/Feature/Anime/View/SettingsViewController.swift:
```diff
@@ -0,0 +1,179 @@
+        audioStackView.addArrangedSubview(audioSwitch)
+        
+        
+        if traitCollection.userInterfaceIdiom == .mac {
```
I don't get why you're doing this? Are you build the app for mac? If you copy code can you just start to think. If you need to actually use the code you're copying. Rather than just adding it in. Reason why is that this looks really bad when you do this in a job or in a tech test cos it shows and means you don't know what is going on. This is massive red flag to employers

tunds commented on Anime-App/Feature/Anime/View/ViewControllers/AnimeViewController.swift:
```diff
@@ -102,12 +114,35 @@ extension AnimeViewController: UISearchBarDelegate {
+        if audioIsOn {
+            print("Sound is playing")
+            audioManager.playSound()
+        } else {
```
So you don't actually need this else block you want to just copy what you did with the haptic feedback where if it's on just play the sound

tunds commented on Anime-App/Feature/Anime/View/ViewControllers/AnimeViewController.swift:
```diff
@@ -102,12 +114,35 @@ extension AnimeViewController: UISearchBarDelegate {
+    
+    @objc func addSound() {
+        let audioIsOn = UserDefaults.standard.bool(forKey: "UseAudio")
+        addHapticFeedBack()
```
This shouldn't be in the function the only thing this should handle is playing sounds, this should be called outside of this function before or after you call the add sounds function you don't want to mix and match them

tunds commented on Anime-App/Feature/Anime/View/ViewControllers/AnimeViewController.swift:
```diff
@@ -102,12 +114,35 @@ extension AnimeViewController: UISearchBarDelegate {
+    }
+    
+    @objc func addSound() {
+        let audioIsOn = UserDefaults.standard.bool(forKey: "UseAudio")
```
Can you move all these user defaults keys to an enum like i said before

tunds commented on Anime-App/Feature/Anime/View/ViewControllers/AnimeViewController.swift:
```diff
@@ -102,12 +114,35 @@ extension AnimeViewController: UISearchBarDelegate {
+        }
+    }
+    
+    @objc func addSound() {
```
A good habit to do is before you play a sound is to stop any sounds that are playing, you don't want to do this in an if statement it should be outside and just the first thing you do. This is to stop sounds from overlapping and playing over one another

tunds commented on Anime-App/Feature/Anime/View/ViewControllers/AnimeViewController.swift:
```diff
@@ -102,12 +114,35 @@ extension AnimeViewController: UISearchBarDelegate {
 extension AnimeViewController {
     
+    func addHapticFeedBack() {
+        let hapticFeedbackIsOn = UserDefaults.standard.bool(forKey: "UseHaptic")
```
Nice this function, is spot on nice work

tunds commented:
Tobi can you look up a tutorial for a settings screen and make this look nicer, right now this doesn't look the best. Use this an inspiration please, lets not just settle for anything and try to build the best thing we can. Once this is done can you update the video with new screenshots and vids of the settings screen

https://www.youtube.com/watch?v=2FigkAlz1Bg

TobiA34 committed:
- Deleted Gundam image
- Added image of me
- Updated audio Manager
- Deleted old setting view controller
- Updated TabBarController
- In AnimeViewController added function to play sound and haptic feedback
- Created model setting option
- Created multiple table view cell for setting screen
- Created setting view model

TobiA34 committed:
Merge branch 'main' into task/settings-screen

TobiA34 committed:
- Added self to spinner.startAnimating
- Added ui view animate in stopAnimating
- Added variable limit and offset in anime view controller
- Added calculation for offset and changed if statement in startPagination method
- Removed spinner in AnimeViewController
